src/suggestEnvironmentVars.ts

```diff
@@ -48,73 +62,107 @@ function createSuggestion(
+    openingBraceIndex >= 0 && closingBraceIndex > openingBraceIndex;
+  return isInside;
+}
 
-        const envVarsObj = getEnvsFromEnvCommand() as Record<
-          string,
-          { src: string; val: string }
-        >; // Explicitly cast to the correct type
-        const envVars = new Map(Object.entries(envVarsObj));
+function getBraceIndicesOfCurLine(
+  currentLine: string,
+  cursorIndex: number
+): [number, number] {
+  console.log("Getting brace indices for current line...");
+  const openingBraceIndex = currentLine.lastIndexOf("${", cursorIndex);
+  const closingBraceIndex = currentLine.indexOf("}", cursorIndex);
+  return [openingBraceIndex, closingBraceIndex];
+}
 
-        const suggestionsArray = Array.from(envVars.entries()).map(
-          ([env, meta]) => createSuggestion(env, meta.val, meta.src, position)
-        );
+function createSuggestionsArray(
+  envVarsObj: EnvVarObject
+): vscode.CompletionItem[] {
+  return Object.entries(envVarsObj).map(([env, data]) => {
+    const completionItem = new vscode.CompletionItem(env);
+    completionItem.documentation = new vscode.MarkdownString()
+      .appendText(`Source: ${data.src}\n`)
+      .appendText(`Value: ${data.val}`);
+    completionItem.detail = "Press Ctrl + Space for docs.";
+    return completionItem;
+  });
+}
 
-        if (triggerPrefix && !triggerPostfix) {
-          return suggestionsArray;
-        } else if (triggerPrefix) {
-          return suggestionsArray;
-        } else {
-          return [];
-        }
-      },
-      ...triggers, // triggers for activating the suggestions
-      resolveCompletionItem(
-        item: vscode.CompletionItem,
-        token: vscode.CancellationToken
-      ) {
-        const editor = vscode.window.activeTextEditor;
-        if (editor) {
-          const position = editor.selection.active;
-          cursorPosition = position.character;
-        }
-        return item;
-      },
+export let lama2ProvideCompletionItems = {
+  provideCompletionItems(
+    document: vscode.TextDocument,
+    position: vscode.Position
+  ) {
+    const currentLine = document.lineAt(position.line).text;
+    const cursorIndex = position.character;
+
+    const [openingBraceIndex, closingBraceIndex] = getBraceIndicesOfCurLine(
+      currentLine,
+      cursorIndex
+    );
+
+    const typedEnvArg = currentLine.substring(
+      openingBraceIndex + 2,
+      cursorIndex
+    );
+
+    const envVarsObj: EnvVarObject = getEnvsFromEnvCommand(typedEnvArg);
+    const isInsidePlaceholder = isCursorInsidePlaceholder(
+      openingBraceIndex,
+      closingBraceIndex
+    );
+
+    if (isInsidePlaceholder) {
+      const completions: vscode.CompletionItem[] =
+        createSuggestionsArray(envVarsObj);
+
+      return completions;
     }
+    return [];
+  },
+};
+
+export function lama2RegisterCompletionItemProvider() {
+  console.log("Setting up ENVs suggestion...");
+  return vscode.languages.registerCompletionItemProvider(
+    "*",
```

Comment:
Does this `*` interfere with other languages/files by any chance? 

Also - does your new logic work if we explicitly specify lama2 here. If not, again a comment is warranted why.

Reply:
If lama2 is given then only our variables will come in suggestions. If * is given everything comes.

Reply:
Got it; please add all comments in the code. The *why* should usually go in the comments.

Reply:
Its not interfering, tried py and js files.